automata/fa/animation.py

```diff
@@ -0,0 +1,376 @@
+from collections import deque
+from collections.abc import Iterable
+from dataclasses import dataclass
+from operator import itemgetter
+from typing import TYPE_CHECKING, Any, Optional
+
+import manim
+
+from automata.base.animation import Animate, _ManimEdge, _ManimInput, _ManimNode
+from automata.base.exceptions import RejectionException
+from automata.fa.fa import FA, FAStateT
+
+if TYPE_CHECKING:
+    from automata.fa.dfa import DFA, DFAStateT
+    from automata.fa.nfa import NFA, NFAStateT
+
+
+class _FAGraph(manim.VDict):
```

Comment:
Is inheritance the right choice here? Or does the rest of the code only call the methods you define within this class? (i.e. the code here doesn't use methods inherited from `manim.VDict`).

Reply:
The only method used that's inherited from `manim.VDict` is the `[]` operator (i.e. `__getitem__`), by using which we get the nodes and edges of the graph. I tried to use a `manim.VDict` object to represents a graph, but later I found that I have to add new method. As a graph can be represented by a dictionary, I think that graph can *be* a dictionary, thus I use inheritance.

Reply:
How frequently is this operator used? Although a graph can be represented with a dictionary, we should avoid inheritance if possible and use composition (i.e. have the `VDict` as a member variable), since that exposes fewer methods to the client code. It would be better to have a couple limited methods instead of using the `[]` operator directly if this is only used in limited ways.

Reply:
> How frequently is this operator used? 

Very Frequently. 😥 There’re lots of occasions where I need to get a node or an edge.

> we should avoid inheritance if possible and use composition (i.e. have the `VDict` as a member variable),

```suggestion
class _FAGraph:
    nodes: manim.VDict
    edges: manim.VDict
```

Would modification like that be OK? In that case, we would use `fa_graph.nodes[node_name]` to get a node and `fa_graph.edges[tail_name, head_name]` to get an edge. Or shall we add a `get_node` method and a `get_edge` method and let the `nodes` and `edges` private? (Though the whole class is already private)

Reply:
I think the change as proposed is also ok. The prior inheritance makes more sense at least since the classes used are smaller, but here I think it's significantly simpler to avoid.